Add direct nexus loading

matthew-d-jones opened the pull request:
Implements `scippneutron.load_nexus()`
Scope in file can be limited using `root` argument, for example to load from specific `NXentry`, or even to load a single `NXlog`. Behaviour in presence of multiple `NXentry` groups etc is documented by unit tests. Unit tests use an in-memory NeXus file which is constructed using a helper class (builder pattern).
If there are loadable event data in the file then a `DataArray` is returned, if there are only logs then a `Dataset` is returned, if nothing is loaded then returns `None`.
Loading is NeXus compliant, in particular it is not sensitive to group names or to the location of data groups in the hierarchy of the file.

I've not optimised performance, but it is currently roughly comparable to loading with `load()` via Mantid. The approach to concatenating data from different event data groups slows things a bit, for example our `PG3_4844_event.nxs` test file contains ~20 event data groups and `load_nexus` takes ~12 seconds, `load` takes ~8 seconds. NB, `load_nexus` does not load pulse times yet, loading pulse times and broadcasting them to events is a foreseeable performance hurdle.

Implemented loading:
- `NXevent_data` - and bin by detector-id
- `NXlog` (with or without `time` dataset)
- pixel positions
- instrument name
- experiment title

Not yet implemented (covered by tickets #19, #20, #21):
- parsing `NXtransformation` chains
- pulse times with event data
- source and sample position
- start of run time
- adjusting log times for `offset` attribute (make consistent, for example from start of run)

SimonHeybrock commented on python/src/scippneutron/_loading_detector_data.py:
```diff
@@ -0,0 +1,201 @@
+    # In other words, ensure that event_index includes the bin edge for
+    # the last pulse.
+    event_id_ds = group["event_id"]
+    event_index = group["event_index"][...].astype(np.int64)
```
Why do we need this to be `int64`? Is it unsigned in the file?

matthew-d-jones replied:
Yes, it is typically uint32, but converting uint32 -> int32 or uint64 -> int64 would probably be better.

SimonHeybrock commented on python/src/scippneutron/_loading_detector_data.py:
```diff
@@ -0,0 +1,201 @@
+    number_of_events = event_index[-1]
+    event_time_offset_ds = group["event_time_offset"]
+    event_time_offset = sc.Variable([_event_dimension],
+                                    values=event_time_offset_ds[...],
+                                    dtype=ensure_not_unsigned(
+                                        event_time_offset_ds.dtype.type),
```
Unless the unsigned->signed conversion is required, can we use `h5py.read_direct` to avoid the extra allocation and copy?

SimonHeybrock commented on python/src/scippneutron/_loading_detector_data.py:
```diff
@@ -0,0 +1,201 @@
+        [_event_dimension], values=event_id_ds[...],
+        dtype=np.int32)  # assume int32 is safe for detector ids
+
+    # Weights are not stored in NeXus, so use 1s
+    weights = sc.Variable([_event_dimension],
+                          values=np.ones(event_id.shape),
+                          dtype=np.float32)
+    data = sc.DataArray(data=weights,
+                        coords={
+                            'tof': event_time_offset,
+                            _detector_dimension: event_id
+                        })
```
There are helpers in `sc.detail`, `move` and `move_to_data_array`, that can avoid copies here (and probably some other places).

SimonHeybrock commented on python/src/scippneutron/_loading_detector_data.py:
```diff
@@ -0,0 +1,201 @@
+        return
+    else:
+
+        def get_detector_id(events_and_max_det_id: DetectorData):
+            # Assume different detector banks do not have
+            # intersecting ranges of detector ids
+            return events_and_max_det_id.detector_ids[0]
```
Would it make sense to sort by the group name in the file?

matthew-d-jones replied:
I don't think so, there is no constraint on naming of detector groups, for DREAM we may have "endcap" and "mantle" for example.

SimonHeybrock commented on python/src/scippneutron/_loading_detector_data.py:
```diff
@@ -0,0 +1,201 @@
+        pixel_positions_loaded = all(
+            [data.pixel_positions is not None for data in event_data])
+        events, _, pixel_positions = astuple(event_data.pop(0))
+        while event_data:
+            new_events, _, new_pixel_positions = astuple(event_data.pop(0))
+            events = sc.concatenate(events, new_events, dim=_event_dimension)
+            if pixel_positions_loaded:
+                pixel_positions = sc.concatenate(pixel_positions,
+                                                 new_pixel_positions,
+                                                 dim=_detector_dimension)
```
So if I understand correctly this is concatenating all event lists before binning into pixels?

Another option would be to *first* do the binning and *then* concatenate. There may be a couple of advantages to this:
- Less memory consumption for temporaries/copies in `sc.bin`.
- Potentially faster `sc.bin` since it would act on pre-grouped data. Lookup of target pixel would only need to consider `Nbank_pixels` rather than `Ninstrument_pixels`.
- More future proof? For example, someone might want to preverse the bank-splitting, e.g., in instruments with detectors for different techniques such as imaging and diffraction, which would not get merged for the workflow?

matthew-d-jones replied:
I've swapped binning and concatenation order as you suggest, and it did improve performance a little.
Another possible approach would be to avoid concatenation entirely by `read_direct` into slices of a single `Variable` for all event data groups. It would be less flexible, but something to consider if the time for the concatenation step is still significant.

SimonHeybrock replied:
I think the real way to improve here is to support concatenate with a vector of inputs, rather than just two inputs.

matthew-d-jones replied:
I thought about that, but wouldn't that involve a large memory overhead?

SimonHeybrock replied:
Why? The end result is the same size, we just avoid all the intermediate steps. In fact the memory used should be less. Say we have banks of sizes A, B, C, D:
- If we go stepwise, largest memory use will be in the last step `ABC + D + ABCD` plus what was previously used for A, B, C, and D individually.
- If we go in one step we just have ABCD.
-

SimonHeybrock commented on python/src/scippneutron/_loading_detector_data.py:
```diff
@@ -0,0 +1,201 @@
+                                                 new_pixel_positions,
+                                                 dim=_detector_dimension)
+
+        # Events in the NeXus file are effectively binned by pulse
+        # (because they are recorded chronologically)
+        # but for reduction it is more useful to bin by detector id
```
There was this issue I had open with ECDC that got recently implemented, to ensure that the file-writter would set a flag when things were *not* ordered by pulse (apprantely this is possible). Can we check for this flag?

matthew-d-jones replied:
It's not useful until we load pulse times though right?
But I can go and check how that was implemented and add some code or comment here to avoid risk of someone else having to go delving in the Filewriter codebase.

SimonHeybrock replied:
Might be useful. I still have in mind that we may want to do certain filtering operations before binning into pixels (e.g., to filter/split by polarization state, or filter out pulses based on bad chopper params). This might be easier if we know it is ordered.

matthew-d-jones replied:
The ticket for flagging unordered data is in selected, not yet implemented, ticket number is ECDC-641.
I'll try to move that along by making a proposal for how the flag could be recorded in file.

SimonHeybrock reviewed (commented):
Only looked at the even-data loading so far:

OwenArnold commented on python/src/scippneutron/_loading_log_data.py:
```diff
@@ -0,0 +1,107 @@
+    try:
+        values = group[value_dataset_name][...]
+    except KeyError:
+        raise BadSource(f"NXlog at {group.name} has no value dataset")
```
I quite like the use of this for encoding compliance with the Nexus standard. Presumably it would be too verbose or too fragile to reference the standard itself?

matthew-d-jones replied:
The problem with that is that everything in the standard is optional. It is not against the standard for the value dataset to be missing, it is simply that we can't do anything with an NXlog that doesn't have a value dataset in our application.

OwenArnold commented on python/tests/test_load_nexus.py:
```diff
@@ -0,0 +1,416 @@
+    expected_counts = np.array([2, 1, 2])
+    assert np.allclose(counts_on_detectors.data.values, expected_counts)
+    expected_detector_ids = np.array([1, 2, 3])
+    assert np.allclose(loaded_data.coords['detector-id'].values,
```
You could use `sc.is_approx` here.

matthew-d-jones replied:
Here I'm comparing against numpy arrays though.

OwenArnold commented on python/tests/nexus_helpers.py:
```diff
@@ -0,0 +1,244 @@
+            time_ds.attrs.create("units", data=log.time_units)
+
+
+class InMemoryNexusFileBuilder:
```
👍

SimonHeybrock commented on python/src/scippneutron/_loading_common.py:
```diff
@@ -0,0 +1,31 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+# Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
```
This is not pointing to the repo but the organization. This is the zero throughout the projects at this point.

matthew-d-jones commented on python/src/scippneutron/_loading_detector_data.py:
```diff
@@ -0,0 +1,201 @@
+        dtype=np.int32)  # assume int32 is safe for detector ids
+
+    # Weights are not stored in NeXus, so use 1s
+    weights = sc.Variable([_event_dimension],
```
@SimonHeybrock now implemented them and I've made use of them.

matthew-d-jones commented:
I think I have addressed all comments so far.
I have not used `read_direct` in a few places, including loading pixel positions, where a flatten operation is required.
The current build failure is the dependency version clash which needs to be resolved in the mantid framework package, I'm looking into that now.

OwenArnold reviewed (commented):
Looks good.

This is one of the first notable features for scippneutron.
1. We should make a mention of this in some release notes. We havent yet created the pages for that. Would be similar to those in scipp.
2. We should maybe exercise the new functionality in some of our other docs, for example https://github.com/scipp/scippneutron/blob/main/docs/tutorials/neutron-diffraction.ipynb

Maybe both of these are best done as followups.

SimonHeybrock commented on python/src/scippneutron/_loading_detector_data.py:
```diff
@@ -0,0 +1,194 @@
+    detector_number_ds_name = "detector_number"
+    if detector_number_ds_name in group.parent:
+        # Hopefully the detector ids are recorded in the file
+        detector_ids = group.parent[detector_number_ds_name][...].flatten()
```
Wondering why we are flattening this and the positions in a file above? Can't we keep the dimensionality?

matthew-d-jones replied:
If there is only one detector then sure, but otherwise it would surely lead to problems when concatenating with other detectors?

SimonHeybrock replied:
You are thinking along the lines of having, e.g., the DREAM mantle and endcap detectors, which have different logical 3D pixel volumes and could not be concatenated?

matthew-d-jones replied:
The DREAM geometry is complicated enough that I just recorded positions and ids in 1D datasets anyway.
But it seems to be usual to record ids and positions for a simple rectangular imaging detector panel in 2D, to mirror the physical layout of pixels. If you have an instrument with one detector bank like that, and another bank with 1D id and position datasets, then there would be a problem.

SimonHeybrock replied:
Maybe it should just fail then? Ignoring the dimensions that are contained in the file feels odd?

matthew-d-jones replied:
The current implementation allows loading all data in all cases, it seems like the right place to start to me. 
Retaining the dimensionality would be possible in conjunction with the option of not concatenating the data from different panels. But that feels like adding complexity for guessed requirements at this stage.

SimonHeybrock replied:
👍

SimonHeybrock commented on python/src/scippneutron/_loading_detector_data.py:
```diff
@@ -0,0 +1,194 @@
+
+    detector_ids = sc.Variable(dims=[_detector_dimension],
+                               values=detector_ids,
+                               dtype=np.int32)
```
While `int32` is what Mantid uses, is there any reason against keeping what the file contains (with a potential switch to signed)?

matthew-d-jones replied:
I've now changed it to retain the original type. This required checking that event ids and detector numbers are the same integer type in order to give a nice warning to the user and skip the current event group. I added units tests to cover this.

SimonHeybrock commented on python/src/scippneutron/_loading_detector_data.py:
```diff
@@ -0,0 +1,194 @@
+
+    if not event_data:
+        return
+    else:
```
Maybe just drop the else, for less confusing spacing and indentation?

matthew-d-jones replied:
done

SimonHeybrock commented on python/src/scippneutron/_loading_detector_data.py:
```diff
@@ -0,0 +1,194 @@
+            new_events = sc.bin(detector_data.events,
+                                groups=[detector_data.detector_ids])
+            if pixel_positions_loaded:
+                new_events.coords['position'] = detector_data.pixel_positions
```
Just for my understanding of Nexus: This implies that the files store a separate array of pixel positions for every `NXevent_data`, rather than a single array?

matthew-d-jones replied:
Yes, each `NXdetector` contains the pixel positions and `NXevent_data` for that detector.
Nesting `NXdetector`s to make this more flexible has been discussed, but there are no examples of that in wild yet.

SimonHeybrock commented on python/src/scippneutron/load_nexus.py:
```diff
@@ -0,0 +1,138 @@
+            (nx_event_data, nx_log, nx_entry, nx_instrument), nexus_file[root])
+
+        if len(groups[nx_entry]) > 1:
+            # We can't sensibly load from multiple NXentry, for example each
+            # could could contain a description of the same detector bank
+            # and lead to problems with clashing detector ids etc
```
I suppose we could consider return a dict of data arrays or a dataset then? Is this something that is used in practice? IF so we can consider supporting it in the future.

matthew-d-jones replied:
The only example I've seen of using multiple `NXentry` in a single file was from a light source facility.

SimonHeybrock commented on python/tests/nexus_helpers.py:
```diff
@@ -0,0 +1,244 @@
+        return h5_object.attrs[attribute_name]
+
+
+def find_by_nx_class(
+        nx_class_names: Tuple[str, ...],
+        root: Union[h5py.File, h5py.Group]) -> Dict[str, List[h5py.Group]]:
+    """
+    Finds groups with requested NX_class in the subtree of root
+
+    Returns a dictionary with NX_class name as the key and list of matching
+    groups as the value
+    """
+    groups_with_requested_nx_class: Dict[str, List[h5py.Group]] = {
+        class_name: []
+        for class_name in nx_class_names
+    }
+
+    def _match_nx_class(_, h5_object):
+        if isinstance(h5_object, h5py.Group):
+            try:
+                nx_class = _get_attr_as_str(h5_object, "NX_class")
+                if nx_class in nx_class_names:
+                    groups_with_requested_nx_class[nx_class].append(h5_object)
+            except KeyError:
+                pass
```
There is a very similar-looking method `_find_by_nx_class` above?

matthew-d-jones replied:
Oops, something went wrong in the refactoring

matthew-d-jones replied:
removed duplicate

SimonHeybrock commented:
> Looks good.
> 
> This is one of the first notable features for scippneutron.
> 
>     1. We should make a mention of this in some release notes. We havent yet created the pages for that. Would be similar to those in scipp.

We have, see https://scipp.github.io/scippneutron/about/release-notes.html. However, I somewhat disagree about adding this to the release note. Personally I'd prefer to try this ourselves for a while before publicizing it.

>     2. We should maybe exercise the new functionality in some of our other docs, for example https://github.com/scipp/scippneutron/blob/main/docs/tutorials/neutron-diffraction.ipynb
> 
> 
> Maybe both of these are best done as followups.

Yes, we should *not* do this in this release.
